drake_ros_systems/src/tf_broadcaster_system.cpp

```diff
@@ -0,0 +1,139 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <drake/common/eigen_types.h>
+#include <drake/geometry/query_object.h>
+#include <drake/geometry/scene_graph_inspector.h>
+
+#include <rclcpp/clock.hpp>
+#include <tf2_ros/transform_broadcaster.h>
+
+#include <memory>
+#include <string>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+
+#include "drake_ros_systems/tf_broadcaster_system.hpp"
+#include "drake_ros_systems/utilities/type_conversion.hpp"
+
+
+namespace drake_ros_systems
+{
+class TfBroadcasterSystem::TfBroadcasterSystemPrivate
+{
+public:
+  std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
+  drake::systems::InputPortIndex graph_query_port_index;
+};
+
+TfBroadcasterSystem::TfBroadcasterSystem(
+  DrakeRosInterface * ros,
+  const std::unordered_set<drake::systems::TriggerType> & publish_triggers,
+  double publish_period)
+: impl_(new TfBroadcasterSystemPrivate())
+{
+  impl_->tf_broadcaster_ = ros->create_tf_broadcaster();
+
+  impl_->graph_query_port_index =
+    this->DeclareAbstractInputPort(
+    "graph_query", drake::Value<drake::geometry::QueryObject<double>>{}).get_index();
+
+  // vvv Mostly copied from LcmPublisherSystem vvv
```

Comment:
I could considering `tf2_ros::TransformBroadcaster` is, for the most part, just a publisher. But I purposefully chose to use the official API. If it's implementation ever changes, this system will remain functional.

Reply:
Ah, that rationale makes sense. Using the official API is preferable in this case then.

Reply:
One argument against using the ROS1 TransformBroadcaster was that it was surprisingly inefficient at batching TF update messages and instead sent each transform individually. If that's still the case, I would strongly favor our own publisher-based implementation that batches them together (at least until fixing the upsteam version to batch).

Reply:
`tf2`'s TransformBroadcaster does provide an API to send a collection of messages at once, so I'll guess that by batching you mean the ability to aggregate transforms sent one by one in a given time window (which is indeed not supported). If that's the case, I'd much rather make a feature contribution upstream than to re-implement it here.

Reply:
Yes, `TransformBroadcaster::sendTransform(const std::vector<geometry_msgs::msg::TransformStamped> & msgtf)` should be good enough for our uses.

Although, I would say in terms of code complexity, `TransformBroadcaster` is literally a publisher with unnecessarily verbose sugar for copying a vector. We should not duplicate code here just because - TF broadcasting is not special, so we should use the `RosPublisherSystem`.

Reply:
Alright. I had to check whether the QoS settings that `tf2_ros` expects were exposed. [They are](https://github.com/ros2/geometry2/blob/ros2/tf2_ros/include/tf2_ros/qos.hpp). We can configure our own publisher with them. We can revisit this if it (ever) breaks.

Reply:
See https://github.com/RobotLocomotion/drake-ros/pull/2/commits/5e711c2e9808eeb9cd18098ba09759e4b3b7b014 for this.